### PR TITLE
app/forkjoin: implement generic forkjoin

### PR DESCRIPTION
**Pull request by corverroos**
Implements a generic forkjoin package.

category: feature
ticket: none

### CODE REVIEW

File: app/forkjoin/forkjoin.go [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+type Results[I, O any] <-chan Result[I, O]
+
+// Result contains the input and resulting output from the work function.
+type Result[I, O any] struct {
+	Input  I
+	Output O
+	Err    error
+}
+
+// Flatten blocks and returns all the outputs when all completed and
```
Comment (xenowits) [inline app/forkjoin/forkjoin.go:56]: nit: Flatten function comment is unclear to me

File: app/forkjoin/forkjoin.go [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+
+	// Fork enqueues inputs, keeping track of how many was enqueued.
+	fork := func(i I) {
+		wg.Add(1)
+		input <- i
+	}
+
+	// Join returns the results channel that will contain all the results in the future.
+	// It also closes the input queue (causing subsequent calls Fork to panic)
+	// It also starts a shutdown goroutine that closes the results channel when processing completed
```
Comment (xenowits) [inline app/forkjoin/forkjoin.go:199]:
```suggestion
	// It also starts a shutdown goroutine that closes the results channel when processing completes.
```

File: app/forkjoin/forkjoin.go [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+//     fork(in) // Note that calling fork AFTER join panics!
+//   }
+//
+//   resultChan := join()
+//   // Either read results from the channel as they appear
+//   for result := range resultChan { ... }
+//   // Or block until all results are complete and flatten
+//   results, firstErr := resultChan.Flatten()
+//
+func New[I, O any](ctx context.Context, work Work[I, O], opts ...Option) (Fork[I], Join[I, O]) {
```
Comment (xenowits) [inline app/forkjoin/forkjoin.go:141]: discussion: i'm wondering why we need the `fork-join` API? where is it intended to be used?

**Review by xenowits — APPROVED** [review]
Looks great!

File: app/forkjoin/forkjoin.go [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+		}
+	}
+
+	if otherErr != nil {
+		return resp, otherErr
+	} else if ctxErr != nil {
+		return resp, ctxErr
+	}
+
+	return resp, nil
```
[the remark this answers is not in the document]
Comment (corverroos) [reply app/forkjoin/forkjoin.go:86]: sure this is the same and more concise, but I feel it is a little clearer the original way

File: app/forkjoin/forkjoin.go (hunk 3 above)
Comment (corverroos) [reply app/forkjoin/forkjoin.go:141]: It is similar to `errgroup.Group` or `sync.WaitGroup`, but it just supports `failfast` and result collection out of the box. It is a pattern we manually write in a lot in tests (in the code in a few places), so I thought we can create a library for that

File: app/forkjoin/forkjoin.go [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+)
+
+const (
+	defaultWorkers  = 8
+	defaultInputBuf = 100
+	defaultFailFast = false
+)
+
+// Fork function enqueues the input to be processed asynchronously.
+// Note Fork may block temporarily while the input buffer is full, see WithInputBuffer.
```
[the remark this answers is not in the document]
Comment (corverroos) [reply app/forkjoin/forkjoin.go:34]: not sure about this

File: app/forkjoin/forkjoin.go (hunk 1 above)
Comment (corverroos) [reply app/forkjoin/forkjoin.go:56]: improved

File: app/forkjoin/forkjoin.go [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+func (r Results[I, O]) Flatten() ([]O, error) {
+	var (
+		ctxErr   error
+		otherErr error
+		resp     []O
+	)
+	for result := range r {
+		resp = append(resp, result.Output)
+
+		if result.Err == nil {
```
[the remark this answers is not in the document]
Comment (corverroos) [reply app/forkjoin/forkjoin.go:70]: sure but it makes the code more readable

File: app/forkjoin/forkjoin.go [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+			continue
+		}
+
+		if errors.Is(result.Err, context.Canceled) && ctxErr == nil {
```
[the remark this answers is not in the document]
Comment (corverroos) [reply app/forkjoin/forkjoin.go:74]: nope, should only set `ctxErr` if not set yet

**Pull request merged**